src/app/js/formats/streamgraph/Streamgraph.js

```diff
@@ -0,0 +1,410 @@
+import React, { PureComponent } from 'react';
+import { connect } from 'react-redux';
+import compose from 'recompose/compose';
+import * as d3 from 'd3';
+
+import { getCssToString, zoomFunction, distinctColors, transformDataIntoMapArray, randomId, getMinMaxValue } from './utils';
+import injectData from '../injectData';
+import exportableToPng from '../exportableToPng';
+
+const styles = {
+    divContainer: {
+        overflow: 'hidden',
+        position: 'relative',
+    },
+    tooltip: {
+        marginTop: 900,
+    },
+    vertical: {
+        marginTop: 60,
+        height: 190,
+        pointerEvents: 'none',
+    },
+    legend: {
+        position: 'relative',
+    },
+    legendItem: {
+        marginTop: 2,
+    },
+    legentItemText: {
+        marginLeft: 5,
+    },
+    legendButton: {
+        height: 15,
+        width: 15,
+    },
+};
+
+class Streamgraph extends PureComponent {
+
+    constructor(props) {
+        super(props);
+        this.state = {
+            width: 800,
+            height: 300
+        }
+
+        this.uniqueId = randomId(10);
+        this.divContainer = "divContainer" + this.uniqueId;
+        this.svgContainer = "svgContainer" + this.uniqueId;
+        this.anchor = "anchor" + this.uniqueId;
+
+        this.xAxisScale;
+        this.yAxisScale;
+        this.xAxis;
+        this.yAxisL;
+        this.yAxisR;
+        this.gx;
+        this.gyr;
+        this.gyl;
+        this.streams;
+
+        //this.draft = draft.bind(this);
+        this.zoomFunction = zoomFunction.bind(this);
+    }
+
+    setGraph() {
+        let [valuesObjectsArray, valuesArray, minDate, maxDate, nameList] = transformDataIntoMapArray(this.props.formatData);
+        const divContainerWidth = document.getElementById(this.divContainer).clientWidth;
+        let { height } = this.state;
+        let width = divContainerWidth;
+        const svgHeight = height;
+        const svgWidth = width;
+
+        const d3DivContainerId = "d3DivContainer" + this.uniqueId;
+        const divContainer = d3.select(this.refs.divContainer)
```

Comment:
Is the variable `this.refs` initialized? I don't see it.
Please, use the React refs API, it would be far more readable.

Reply:
For the refs, I tryed to set up using the style `<div ref={this.myRef} />` but d3js cannot append on it through a `d3.select(this.myRef).append("div")`. I have any idea why, searched a lot on the web and only find examples with the `<div ref="myRef" />` style.. Ideas ?

Reply:
@youlot You have to set one ref per element you want to reach:

```jsx
class MyComponent extends Component {
    contructor(props) {
        super(props);
        this.containerRef = React.createRef();
        this.svgRef = React.createRef();
        this.anchorRef = React.createRef();
    }

    render() {
        const { width, height } = this.state;
        return (
            <div id="divContainer" ref="this.containerRef" style={styles.divContainer}>
                <svg id="svgContainer" ref="this.svgRef" width={width} height={height}>
                    <g id="anchor" ref="this.anchorRef" />
                </svg>
            </div>
        );
    }
}
```

Then, [as explained on the doc](https://reactjs.org/docs/refs-and-the-dom.html#accessing-refs), in order to access the DOM element pointed by the ref, you have to write:

```js
console.log(this.containerRef.current);
// So, the following should work:
d3.select(this.containerRef.current).append("div")
```

If needed, here is the API of a DOM element (the `current` part of the ref): https://developer.mozilla.org/en-US/docs/Web/API/element

Reply:
@Kmaschta Yes I tried this method as described in the documentation, one ref per element e.g: `this.containerRef`, initialized in constructor with `React.createRef();` and assigned to an html element using `ref={this.containerRef}`. This works well for using the ref from react but I have any idea why d3js cannot use it, when I try a select() using this ref it just doesn't work as if the ref was broken.

I finally figured it out by using callbacks refs : ` ref={divContainer => this.divContainer = divContainer}` doing this way the ref is good from react side and d3js can access to this ref ! I don't understand yet why this method works while the previous one doesn't but.. it works !

Reply:
The refs callback is the legacy way of doing so. It's not optimal and will be deprecated soon.

Did you try with the current key ? `this.containerRef.current` when you selected?

Reply:
Are you sure the callback refs will be deprecated ? I haven't seen anything about it in the [documentation](https://reactjs.org/docs/refs-and-the-dom.html). I have seen that the string ref will be removed in one of the future releases and that's why they recommend using the callback pattern or the createRef api.

I tried the .current way in association with the string ref assignment and it didn't worked but doing the assignment this way : `ref={this.divContainer}` works with the .current ! I will commit this usage because I think that's more appropriate.

Reply:
I'm not fully certain that it'll be deprecated, but indeed, the second option is best! Good job on that :1st_place_medal:

Reply:
Okay, thanks for your help ! :)